modvendor/modvendor.go

```diff
@@ -0,0 +1,223 @@
+		Str("url", src.URL).
+		Str("path", src.Path).
+		Str("ref", src.Ref).
+		Logger()
+
+	if src.Ref == "" {
+		// TODO(katcipis): handle default references.
+		// for now always explicit is fine.
+		return "", errors.E("src %v reference must be non-empty", src)
+	}
+
+	// TODO(katcipis): test that if vendor contains path with matching ref
+	// it will do nothing.
+
+	logger.Trace().Msg("setting up tmp workdir")
+
+	workdir, err := os.MkdirTemp("", "terramate-vendor")
+	if err != nil {
+		return "", errors.E(err, "creating workdir")
+	}
+	defer func() {
+		// We ignore the error here since after the final os.Rename
+		// the workdir will be moved and won't exist.
+		_ = os.Remove(workdir)
+	}()
+
+	clonedir := filepath.Join(vendordir, src.Path, src.Ref)
+
+	logger = logger.With().
+		Str("workdir", workdir).
+		Str("clonedir", clonedir).
+		Logger()
+
+	logger.Trace().Msg("setting up git wrapper")
+
+	g, err := git.WithConfig(git.Config{
+		WorkingDir: workdir,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	logger.Trace().Msg("cloning to workdir")
+
+	if err := g.Clone(src.URL, workdir); err != nil {
+		return "", err
+	}
+
+	// TODO(katcipis): remove .git before moving.
+
+	if err := os.MkdirAll(filepath.Dir(clonedir), 0775); err != nil {
+		return "", errors.E(err, "creating mod dir inside vendor")
+	}
+
+	logger.Trace().Msg("moving cloned mod from workdir to clonedir")
+	if err := os.Rename(workdir, clonedir); err != nil {
+		// This may leave intermediary created dirs hanging on vendordir
+		// since we just create all and then delete clone dir on a failure to move.
+		// If we get a lot of errors from os.Rename we may need to handle this
+		// more gracefully, here we assume that os.Rename errors are rare since both
+		// dirs where just created.
+		errs := errors.L()
+		errs.Append(errors.E(err, "moving cloned module"))
+		errs.Append(os.Remove(clonedir))
+		return "", errs.AsError()
+	}
+	return clonedir, nil
+}
+
+// ParseSource parses the given modsource string.
+// The modsource must be a valid Terraform Git/Github source reference as documented in:
+//
+// - https://www.terraform.io/language/modules/sources
+//
+// Source references that are not Git/Github are not supported.
+func ParseSource(modsource string) (Source, error) {
+	ref := ""
+	splitParams := strings.Split(modsource, "?")
+	modsource = splitParams[0]
+
+	if len(splitParams) > 1 {
+		if len(splitParams) != 2 {
+			return Source{}, errors.E(ErrInvalidModSrc, "unexpected extra '?' on source")
+		}
+
+		refParam := splitParams[1]
+		splitRefParam := strings.Split(refParam, "=")
+
+		if len(splitRefParam) != 2 {
+			return Source{}, errors.E(ErrInvalidModSrc, "parsing ref param %q", refParam)
+		}
+
+		if splitRefParam[0] != "ref" {
+			return Source{}, errors.E(ErrInvalidModSrc, "unknown param %q", splitRefParam[0])
+		}
+
+		ref = splitRefParam[1]
+		if ref == "" {
+			return Source{}, errors.E(ErrInvalidModSrc, "ref param %q is empty", refParam)
+		}
+	}
+
+	switch {
```

Comment:
`source` most commonly is a relative path but it's not handled here. Is the caller always ensuring this is a git/remote path?

Reply:
when the source doesn't match any of the git patterns we generated an unsupported source error, which is the same when we have hg:// etc. Why this would be a problem particularly to local files ? (they will be ignored in the end)

Reply:
fair enough. 
I see that this function can be reused in the `tf.ParseModules()` and then it could have a `Source.Scheme` and a `source.IsLocalPath()` or something.

Reply:
fair enough. 
I see that this function can be reused in the `tf.ParseModules()` and then it could have a `Source.Scheme` and a `source.IsLocalPath()` or something.

Reply:
Ah...hmm it could be useful to generalize it but I'm not focusing on that right now, sorry :heart: